src/lib/rust.ts

```diff
@@ -0,0 +1,261 @@
+import { replaceHomeDir } from "./helpers/replaceHomeDir";
+import { unparseFlags } from "./helpers/unparseFlags";
+import {
+  ensureUserFluenceCargoCratesPath,
+  ensureUserFluenceCargoDir,
+} from "./paths";
+
+const CARGO = "cargo";
+const RUSTUP = "rustup";
+
+const ensureRust = async (commandObj: CommandObj): Promise<void> => {
+  if (!(await isRustInstalled())) {
+    if (commandObj.config.windows) {
+      commandObj.error(
+        "Rust needs to be installed. Please visit https://www.rust-lang.org/tools/install for installation instructions"
+      );
+    }
+
+    const rustupInitFlags = unparseFlags(
+      {
+        quiet: true,
+        y: true,
+      },
+      commandObj
+    );
+
+    await execPromise(
+      `curl --proto '=https' --tlsv1.2 -sSf https://sh.rustup.rs | sh -s -- ${rustupInitFlags}`,
+      "Installing rust"
+    );
+
+    if (!(await isRustInstalled())) {
+      commandObj.error(
+        `Installed rust without errors but ${color.yellow(
+          RUSTUP
+        )} or ${color.yellow(CARGO)} not in PATH`
+      );
+    }
+  }
+
+  if (!(await hasRequiredRustToolchain())) {
+    await execPromise(
+      `${RUSTUP} install ${RUST_TOOLCHAIN_REQUIRED_TO_INSTALL_MARINE}`,
+      `Installing ${color.yellow(
+        RUST_TOOLCHAIN_REQUIRED_TO_INSTALL_MARINE
+      )} rust toolchain`
+    );
+    if (!(await hasRequiredRustToolchain())) {
+      commandObj.error(
+        `Not able to install ${color.yellow(
+          RUST_TOOLCHAIN_REQUIRED_TO_INSTALL_MARINE
+        )} rust toolchain`
+      );
+    }
+  }
+
+  if (!(await hasRequiredRustTarget())) {
+    await execPromise(
+      `${RUSTUP} target add ${RUST_WASM32_WASI_TARGET}`,
+      `Adding ${color.yellow(RUST_WASM32_WASI_TARGET)} rust target`
+    );
+    if (!(await hasRequiredRustTarget())) {
+      commandObj.error(
+        `Not able to install ${color.yellow(
+          RUST_WASM32_WASI_TARGET
+        )} rust target`
+      );
+    }
+  }
+};
+
+const isRustInstalled = async (): Promise<boolean> => {
+  try {
+    await execPromise(`${CARGO} --version`);
+    await execPromise(`${RUSTUP} --version`);
+    return true;
+  } catch {
+    return false;
+  }
+};
+
+const hasRequiredRustToolchain = async (): Promise<boolean> =>
+  (await execPromise(`${RUSTUP} toolchain list`)).includes(
+    RUST_TOOLCHAIN_REQUIRED_TO_INSTALL_MARINE
+  );
+
+const hasRequiredRustTarget = async (): Promise<boolean> =>
+  (await execPromise(`${RUSTUP} target list`)).includes(
+    `${RUST_WASM32_WASI_TARGET} (installed)`
+  );
+
+const cargoInstall = async ({
+  packageName,
+  version,
+  isNightly,
+  isGlobalDependency,
+  commandObj,
+  message,
+}: CargoDependencyInfo & {
+  version: string;
+  commandObj: CommandObj;
+  message: string;
+}): Promise<string> =>
+  execPromise(
+    `${CARGO}${
+      isNightly === true ? " +nightly" : ""
+    } install ${packageName} ${unparseFlags(
+      {
+        version,
+        ...(isGlobalDependency === true
```

Comment:
Because I think type coercion in JS is a great source of bugs. So I disallow it in a way with linter rules. isGlobalDependency parameter can be undefined, and we know undefined is falsy, but depending on the context we might want to do different stuff depending on whether isGlobalDependency is undefined or false. In this case we don't care, so we explicitly write that we care only about `true` case